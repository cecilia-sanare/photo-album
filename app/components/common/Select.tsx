import * as styles from './Select.module.scss';
import { useCachedState } from '../../hooks/use-cached-state';

export type SelectItem = {
  label: string;
  value: string;
}

type SelectProps = {
  items: SelectItem[];
  value?: string;
  placeholder?: string;
  onChange?: (value: string) => void;
}

export function Select({
  items,
  value,
  placeholder,
  onChange
}: SelectProps) {
  const [selectedValue, setSelectedValue] = useCachedState(() => value || '', [value]);

  return (
    <select
      className={styles.select}
      data-testid='select'
      value={selectedValue}
      onChange={(event) => {
        setSelectedValue(event.target.value);
        onChange?.(event.target.value);
      }}
    >
      <option value='' disabled>{placeholder}</option>
      {items.map((item) => (
        <option key={item.value} value={item.value}>{item.label}</option>
      ))}
    </select>
  )
}
