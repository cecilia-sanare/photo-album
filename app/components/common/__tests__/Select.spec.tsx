import { fireEvent, render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Select, SelectItem } from '../Select';
import Chance from 'chance';

const chance = new Chance();

describe('<Select />', () => {
  it('should render the Select', () => {
    const component = render(<Select items={[]}/>);

    expect(component.baseElement).toMatchSnapshot();
  });

  describe('prop(value)', () => {
    it('should select the matching item', () => {
      const expectedItem: SelectItem<string> = {
        label: chance.string(),
        value: chance.string()
      };
      const component = render(<Select items={[expectedItem]} value={expectedItem.value} />);

      expect(component.getByDisplayValue(expectedItem.label)).toBeTruthy();
    });
  });

  describe('prop(onChange)', () => {
    it('should support providing a callback', async () => {
      const onChange = jest.fn();
      const expectedItem: SelectItem<string> = {
        label: chance.word(),
        value: chance.word()
      };

      const component = render(<Select items={[expectedItem]} onChange={onChange} />);

      expect(onChange).toHaveBeenCalledTimes(0);

      await userEvent.selectOptions(component.getByTestId('select'), expectedItem.value);

      expect(onChange).toHaveBeenCalledTimes(1);
    });

    it('should support not providing a callback', async () => {
      const expectedItem: SelectItem<string> = {
        label: chance.word(),
        value: chance.word()
      };

      const component = render(<Select items={[expectedItem]} />);

      expect(component.queryByDisplayValue(expectedItem.label)).toBeFalsy();

      await userEvent.selectOptions(component.getByTestId('select'), expectedItem.value);

      expect(component.getByDisplayValue(expectedItem.label)).toBeTruthy();
    });
  });
});
