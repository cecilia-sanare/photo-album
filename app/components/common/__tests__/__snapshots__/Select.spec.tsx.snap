// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select /> should render the Select 1`] = `
<body>
  <div>
    <select
      class="select"
      data-testid="select"
    >
      <option
        disabled=""
        value=""
      />
    </select>
  </div>
</body>
`;
